### ISSUE TITLE
wxPython 4.0.4/wx.Accessible: generic labels seen by screen readers if the control elects not to implement IAccessible objects

### ISSUE THREAD

**Issue by josephsl** [wxWidgets/Phoenix#1129]
**Operating system**:
Windows 10 Version 1809 (build 17763)

**wxPython version & source**:          
4.0.4/wxWidgets 3.0.5, obtained via PIP

**Python version & source**:            
Python 2.7 32-bit

**Description of the problem**:
If a wxWidgets control (wx.Control) elects not to implement IAccessible objects interface for the underlying window, screen readers such as Narrator and Nonvisual Desktop Access will see a generic control label.

Steps to reproduce:

1. Obtain Nonvisual Desktop Access (NVDA) source via https://github.com/nvaccess/nvda (use Git to clone this repo, passing in --recursive option to fetch submodules including wxPython).
2. Compile NVDA from source by following the directions from the readme file.
3. With Python 2.7 installed, start NVDA.
4. When NVDA starts, the welcome dialog should be displayed. Press Tab to move around this window.
5. Press Control+Windows=Enter to start Narrator.
6. Try navigating NVDA's welcome dialog with Narrator by pressing Tab key.

Expected: screen readers announces proper label for Welcome dialog controls, such as "keyboard layout".
Actual: generic control role labels are announced by NVDA and Narrator, such as "choice" instead of "keyboard layout". The labels are shown on screen as expected.

**Behavior comparisons between 4.0.3 and 4.0.4**:
In 4.0.3, NVDA and Narrator announces correct labels for controls as returned by MSAA itself. However, in 4.0.4, possibly due to introduction of wx.Accessible or something else, it appears wx.Control objects might consult MSAA objects to return labels, and if MSAA objects (server side) are not defined for a control, it might fall back to setting generic control role text as labels. This is reflected in MSAA objects as screen readers see it - in case of NVDA, IAccessible object's accName property returns the generic label in 4.0.4 versus correct label in 4.0.3.

**Impact**:
This is a severe regression for NVDA, as it is a wxWidgets application, and an annoyance for Narrator users and developers (Microsoft) who may encounter apps written in wxPython 4.0.4 in the future. The ideal solution is to let apps specify if they wish to use native versus wxWidgets MSAA server implementation. Although I and some NVDA developers acknowledge that the problem may lie with NVDA source code, the fact that this problem was seen with two screen readers (NvDA and Narrator) indicates that sometimes, a feature that might open up doors for accessible and usable apps might not work in all cases. For NVDA developers, this issue may come up when discussing pros and cons of using newer wxPython releases.

**Relevant NVDA issue**:
An issue for NVDA has been created at nvaccess/nvda:9130.

Thanks.

**Comment by josephsl** [comment]
Hi,

This issue was identified as a wxPython issue by another NVDA contributor.

CC @MichaelDCurran, @LeonardDer, @feerrenrut, @JCSTeh, @PPatel

**Comment by leonardder** [comment]
Additional research reveals that setting name to an empty string at control construction time restores the old behaviour.

The [wx.Window documentation](https://wxpython.org/Phoenix/docs/html/wx.Window.html) says in regard to GetName and SetName:

> Note: This name is not guaranteed to be unique; it is up to the programmer to supply an appropriate name in the window constructor or via SetName

If it is up to the programmer, what is an appropriate name? While documentation says that SetName sets the window's name, it is not mentioned that this name is also revealed to assistive technologies, such as screen readers. My initial thought when I read the description of the Name property years ago, was that it was meant to be a descriptive name for the control that made sense to programmers only, similar to the name of the variable you might assign a Window's instance to.

In my opinion, the ideal solution would be decoupling IAccessible accName from the wx.Window.Name property. There are two ways I can think of for people to provide an accessible name override if necessary: 

1. Create a wx.Accessible and set the accessible name using that object.
2. Similar to Windows Forms: provide an AccessibleName property on wx.Window that can also be set using a constructor keyword argument. Note that Windows Forms also has properties for AccessibleDescription and AccessibleRole.

**Comment by leonardder** [comment]
cc @RobinD42

**Comment by RobinD42** [comment]
wxPython doesn't do anything special here, it just wraps wxWidgets. So any changes will need to be made there.  Perhaps your expectations for how things should work just do not match up with how they are designed to work?

Please create a ticket at https://trac.wxwidgets.org/ or ask questions in the wx-users google group.

**Comment by leonardder** [comment]
> wxPython doesn't do anything special here, it just wraps wxWidgets. So any changes will need to be made there. Perhaps your expectations for how things should work just do not match up with how they are designed to work?

Ah, thanks for the clarification. Yet, upgrading wxPython from 4.0.3 to 4.0.4 and above causes a major decrease in out of the box accessibility.
If I'm correct, the default version of wx has wxUSE_ACCESSIBILITY set to 0, so people have to manually enable it. In that case, I would certainly advise to disable accessibility support by default for wxPython until there's a proper solution in wxWidgets.

**Comment by leonardder** [comment]
From include/wx/msw/setup.h of version 3.0.4:

```
// Use wxAccessible for enhanced and customisable accessibility.
// Depends on wxUSE_OLE.
//
// Default is 0.
//
// Recommended setting (at present): 0
#define wxUSE_ACCESSIBILITY 1
```

I will build WX 3.0.4 stable as well as wx 3.1.2 to see what the accessibility of the samples look like, and will report back my findings.

**Comment by josephsl** [comment]
Hi,

Update: no longer reproducible in wxPython 4.1.0-dev (powered by wxWidgets master/3.1.x) - that is, controls are correctly labeled. This was confirmed with Python 2.7 and 3.7.

I'm closing this for now. Please reopen (with a different title) if the problem is seen in 4.1.0 stable/beta (likely later this year).

Thanks.

**Comment by josephsl** [comment]
Follow-up: let me close the related issue on NVDA issue tracker, too.

Thanks.

**Comment by mush42** [comment]
Hi @josephsl, have you tested the 4.4.1+ versions with Windows 7 (32/64 bit). It seams that the problem still persists.

**Comment by josephsl** [comment]
Hi, you mean 4.0.4? In theory, 4.1.0 should give no problems. Thanks.
